### DISCUSSION TITLE
KeyReleaseEvent intent of keyBase and keyModified

### DISCUSSION THREAD

**Discussion by kannode** [vsg-dev/VulkanSceneGraph#661] [General]
Hi,

I am debugging Win32 keyboard-related issues.
What is the intent of keyBase and keyModified ?

I assume keyBase is more or less the scan_code, whereas keyModified is the final semantic keypress such as vsg::KEY_F1 or vsg::KEY_p ?

Sincerely,
Suresh.

**Comment by robertosfield** [comment]
The keyModified is the the key symbol with shift/ctrl/alt/caps lock applied, the keyBase is the unmodified key value.

To keep things consistent across platforms the keyBase isn't always the raw scan code, that's the intent at least, I don't have a Windows system myself, and didn't write the Windows code so can't verify that.

**Reply by lufriem** [reply]
I'm sorry to be probably a bit dense about this. I'm not sure what the expected behaviour is.. I'll sketch a few situations:

Note: keyModifier is not working on Windows right now, see issue #919. So I'm not really sure what values keyModifier should hold at times..

- suppose the user presses and releases the 'n' key (no capslock, no numlock). This results in

| event | keyBase | keyModified | keyModifier |
|---|---|---|---|
| keyPress | KEY_n | KEY_n | MODKEY_OFF|
| keyRelease | KEY_n | KEY_n | MODKEY_OFF|

So far, so good.

- suppose the user presses the 'ShiftL+n' key (again, no capslock/numlock), by first pressing (and holding down) 'ShiftL', then the 'n' key.

| event | keyBase | keyModified | keyModifier | comment |
|---|---|---|---|---|
| keyPress | KEY_Shift_L| KEY_Shift_L| MODKEY_Shift|
| keyPress | KEY_n | key_N| MODKEY_OFF| this is issue #919|

So the keyModified is upper-case N because of the 'shift' modifier.

-  now the user releases the 'n' key, then the 'ShiftL' key

| event | keyBase | keyModified | keyModifier | comment|
|---|---|---|---|---|
| keyRelease | KEY_n | key_N| MODKEY_OFF| #919 ??|
| keyRelease | KEY_Shift_L| KEY_Shift_L| MODKEY_Shift|

I think that's OK too (but for the #919 issue on Windows)

- Where things get interesting is if the user presses ShiftL, then 'n' and releases the 'Shift' key first..

| event | keyBase | keyModified | keyModifier |comment|
|---|---|---|---|---|
| keyPress | KEY_Shift_L| KEY_Shift_L| MODKEY_Shift|
| keyPress | KEY_n| key_N| MODKEY_OFF  |#919|
| keyRelease | KEY_Shift_L| KEY_Shift_L| KEY_Shift_L| his one is hard to see in the UI, very quick|
| keyPress | KEY_n | KEY_n | MODKEY_OFF | where is this coming from?|
| keyRelease | KEY_n | KEY_n | MODKEY_OFF |

For me, on Windows, using the vsginput sample, an additional 'keyPress' event seems (sometimes?) to be inserted here. I an not sure what the 'Shift' keyRelease should look like, shouldn't this somewhow indicate that the 'n' key is still being pressed?

- Another interesting thing is if I press 'n' first, then 'ShiftL'

| event | keyBase | keyModified | keyModifier |
|---|---|---|---|
| keyPress | KEY_n| KEY_n| MODKEY_OFF |
| keyPress | KEY_Shift_L| KEY_Shift_L| MODKEY_Shift|

In this case I'd expect the second keypress to somehow indicate that the 'n' key was down? Otherwise, how could the application react to a shortcut with this key combination? Are applications supposed to keep track if the keys that are being pressed?
